rclcpp/include/rclcpp/executor.hpp

```diff
@@ -76,9 +77,124 @@ class Executor
+   * This is a helper function that allows a user to pass multiple nodes with callback groups in
+   * one call.
+   * \param[in] node_to_groups a map data structure that maps a node shared ptr to a vector of
+   * callback group shared pointers.
+   * \param[in] notify True to trigger the interrupt guard condition during this function. If
+   * the executor is blocked at the rmw layer while waiting for work and it is notified that a new
+   * callback group was added, it will wake up.
+   */
+  RCLCPP_PUBLIC
+  virtual void
+  add_callback_groups(
+    std::map<rclcpp::node_interfaces::NodeBaseInterface::SharedPtr,
+    std::vector<rclcpp::CallbackGroup::SharedPtr>> node_to_groups,
+    bool notify = true);
+
+  /// Get callback groups that belong to executor
+  /**
+   * This function returns a vector of weak pointers that point to callback groups that were
+   * associated with the executor. The callback groups associated might have been added with
+   * `add_callback_groups`, added when a node is added to the executor with `add_node`, or
+   * a callback group that is not associated to an executor and is in an 'allowable' state
+   * from a node that has been added to an executor.
+   * \return a vector of weak pointers that point to callback groups that are associated with
+   * the executor
+   */
+  RCLCPP_PUBLIC
+  virtual std::vector<rclcpp::CallbackGroup::WeakPtr>
+  get_callback_groups();
+
+  /// Remove a callback group from the executor.
+  /**
+   * The callback group is removed from the map `weak_groups_to_nodes` and unassociated
+   * with the executor. If the node that is associated with the callback group removed
+   * no longer has any callback groups associated with the executor, the guard
+   * condition is removed from `weak_nodes_to_guard_conditions`.
+   * \param[in] group_ptr Shared pointer to the callback group to be added.
+   * \param[in] notify True to trigger the interrupt guard condition during this function. If
+   * the executor is blocked at the rmw layer while waiting for work and it is notified that a
+   * callback group was removed, it will wake up.
+   */
+  RCLCPP_PUBLIC
+  virtual void
+  remove_callback_group(
+    rclcpp::CallbackGroup::SharedPtr group_ptr,
+    bool notify = true);
+
   /// Add a node to the executor.
   /**
    * An executor can have zero or more nodes which provide work during `spin` functions.
+   * If a node is already associated with an executor, the executor throws an exception.
+   * The node is added to `weak_nodes_` which will be iterated before `collect_entities`
+   * from the memory strategy to find any callback groups that are not associated to
+   * an executor and are in an `allowable` state. All callback groups found in the node
+   * the moment it was added to the executor that are not associated to an executor and
+   * in an 'allowable' state will be added to the executor through `add_callback_group`.
```

Comment:
You should mention here what happens to callback groups added to the node later.